### ISSUE TITLE
Global caching of Reflection

### ISSUE THREAD

**Issue by Odonno** [microsoft/dotnet#587]
I am not sure if it is already done by .NET natively or by a Library but I tihkn it could be a great addition to the Framework to enable caching of Reflection info during the runtime of the application. I mean, it can drastically improve the performance of an app which use Reflection on the same types a lot.

```csharp
ReflectionCache.Enable(); // To enable global caching
ReflectionCache.Disable(); // To disable global caching
```

There could be a single method to enable/disable caching and everything (caching) is done as a black box. Each time a method like `GetType()` or `GetProperties()` is called, the data will be retrieved either from the existing method if no cache or from the cache directly.

**Comment by richlander** [comment]
Interesting idea. You could build this yourself. I'd be curious if you did and reported the win. We're unlikely to go the pure back box route but a reflectioncache route would be interesting on top.

**Comment by jnm2** [comment]
Black box would be necessary if you want it to affect, say, Entity Framework.

**Comment by Odonno** [comment]
@jnm2 Indeed. My first thought was to cache everything when a user/library calls a method in the Reflection namespace. So .NET code should be updated to handle that.

Otherwise, the alternative would be to create a library on top of Reflection which either cache or use the Reflection method by (re)writing the same properties & methods you can find in the Reflection namespace.

**Comment by jnm2** [comment]
Reminds me of https://github.com/mgravell/fast-member#fast-access-to-net-fieldsproperties.

**Comment by Odonno** [comment]
Yep, kind of. But it should support constructors and methods too.

**Comment by richlander** [comment]
Do you have any data on what you think the performance win would be given this product change? I'm assuming you are a user of some reflection-heavy scenarios and see this change as a significant performance win for those scenarios. Is that right?

**Comment by Odonno** [comment]
@richlander I am currently working with Reflection on 2 occasions : 

1. With the SuccincT library in which I added a new extension called `With()` which mimic the behavior of the same method in F#. In some cases, I can call this extension a lot of time on the same types. That's why I added more code around it to store the Type information in a Dictionary so the (2 + N) times you want to copy an object, in the attempt that the time required to check type information to be way faster than the first call. I got the idea to make a global cache from this addition to the library, and in the hope that maybe it can be effortless in the future to cache Reflection information.

2. With some side projects where I need to check Type properties. And again sometimes a lot of calls to reflection can be made.

I did not made a benchmark to see how much faster it is with a cached Dictionary. Will try to do it when I can.

**Comment by kalaskarsanket** [comment]
As we are still waiting on you to try out the recommendation suggested, I am going to go ahead and close this issue. Kindly raise a new issue in the appropriate repo as mentioned in [dotnet/1275](https://github.com/microsoft/dotnet/issues/1275)

**Comment by Odonno** [comment]
Wow, that was like 170 years. :) 

@kalaskarsanket No problem. I think we all evolved from there.
